Add ReentrancyGuard status getter

zhiqiangxu opened the pull request:
This PR adds a entered modifier to ReentrancyGuard contract, it's useful for callbacks.

Amxx commented:
Hello @zhiqiangxu 

I'm not sure about this one. If multiple functions in the contract are nonReentrant, you don't know which one was you are entering from. You are also not sure what path you are entering trough, and if you are not entering twice (either one after the other or one within the other).

Can you give us more details about your usecase?

zhiqiangxu commented:
> Hello @zhiqiangxu
> 
> I'm not sure about this one. If multiple functions in the contract are nonReentrant, you don't know which one was you are entering from. You are also not sure what path you are entering trough, and if you are not entering twice (either one after the other or one within the other).
> 
> Can you give us more details about your usecase?

Your concern makes sense, I now prefer just exposing the guard status and let users decide his own logic?

frangio commented:
I'm ok with exposing the guard status but we shouldn't make this function virtual.

zhiqiangxu commented:
> I'm ok with exposing the guard status but we shouldn't make this function virtual.

Yeah internal is enough.

frangio commented:
Coverage wants you to add a test for the new getter.

frangio commented on contracts/security/ReentrancyGuard.sol:
```diff
@@ -66,4 +66,11 @@ abstract contract ReentrancyGuard {
+    /**
+     * expose _status as internal function, the caller decides what to do with it
+     */
+    function _guardStatus() internal view returns (uint256) {
```
Let's call this `_reentrancyGuardStatus`?

zhiqiangxu replied:
Done.

zhiqiangxu commented:
> Coverage wants you to add a test for the new getter.

Added a check in an existing nonReentrant function, let's see if Coverage will be happy this time.

frangio reviewed (dismissed):
Looks good to me! Awaiting review by @Amxx.

frangio commented on contracts/mocks/ReentrancyMock.sol:
```diff
@@ -35,6 +35,8 @@ contract ReentrancyMock is ReentrancyGuard {
         bytes4 func = bytes4(keccak256("callback()"));
         attacker.callSender(func);
+
+        require(_reentrancyGuardStatus() == 2, "ReentrancyMock: _reentrancyGuardStatus failed");
```
I don't think this is getting executed, the attacker line above probably reverts.

Please add a separate test specifically for the guard status, and for the various values it can have.

Amxx replied:
+1 for specific testing

zhiqiangxu replied:
Done. 

Sorry for late response, busy doing a bunch of other things lol:)

frangio reviewed (commented):
Realized tests still need work. The coverage report is clearly showing that the getter is not tested.

![](https://user-images.githubusercontent.com/481465/192562736-58002ae0-3a7f-4e31-ac81-f73497e67b78.png)

Amxx reviewed (changes requested):
Looks mostly good. Need better testing.

I'd do an function that emit the value of the guard ... and this function would be called through 2 public function (one with the modifier, one without)

frangio commented:
Sorry for the back and forth. I realized the getter was leaking the exact values we use for entered and not entered, and these are values we might want to change in the future, given that they were chosen to optimize gas usage and the optimal values may be different once again in the future.

So in order to avoid leaking the exact values I've changed the function to return a boolean and renamed the function accordingly.

@zhiqiangxu @Amxx Let me know what you think.

zhiqiangxu commented:
> Sorry for the back and forth. I realized the getter was leaking the exact values we use for entered and not entered, and these are values we might want to change in the future, given that they were chosen to optimize gas usage and the optimal values may be different once again in the future.
> 
> So in order to avoid leaking the exact values I've changed the function to return a boolean and renamed the function accordingly.
> 
> @zhiqiangxu @Amxx Let me know what you think.

Cool, I think that's better !

zhiqiangxu commented:
Any update?
